src/autorest-core/lib/pipeline/pipeline.ts

```diff
@@ -94,9 +102,130 @@ function CreateCommonmarkProcessor(): PipelinePlugin {
     }
   };
 }
+function CreateArtifactEmitter(inputOverride?: () => Promise<DataStoreViewReadonly>): PipelinePlugin {
+  return async (config, input, working, output) => {
+    return EmitArtifacts(
+      config,
+      config.GetEntry("input-artifact" as any),
+      key => ResolveUri(
+        config.OutputFolderUri,
+        safeEval<string>(config.GetEntry("output-uri-expr" as any), { $key: key, $config: config.Raw })),
+      inputOverride ? await inputOverride() : input,
+      config.GetEntry("is-object" as any));
+  };
+}
 
-export async function RunPipeline(config: ConfigurationView, fileSystem: IFileSystem): Promise<void> {
-  const barrier = new OutstandingTaskAwaiter();
+function* WithIndex<T>(collection: Iterable<T>): Iterable<[T, number]> {
+  let idx = 0;
+  for (const x of collection) {
+    yield [x, idx++];
+  }
+}
+
+function BuildPipeline(config: ConfigurationView): { pipeline: { [name: string]: PipelineNode }, configs: { [jsonPath: string]: ConfigurationView } } {
+  const cfgPipeline = config.GetEntry("pipeline" as any);
```

Comment:
derive pipeline graph entirely from configuration!
Will add more and more convenience into that, currently one would have to rewire the graph to add an own pipeline step.
Example: Inserting a Go simplifier
``` yaml
pipeline:
  # new step:
  go/simplify:
    plugin: go-simplifier
    input: transform # formerly consumed by go/emitter
    output-artifact: source-file-go
  # rewire emitter to consume the above
  go/emitter:
    input: simplify # was: transform
    scope: scope-go/emitter
```